File: CS/ViewModels/MessengerViewModel.cs
namespace DXHtmlMessengerSample.ViewModels {
    using System.Threading.Tasks;
    using DevExpress.DevAV.Chat;
    using DevExpress.DevAV.Chat.Commands;
    using DevExpress.DevAV.Chat.Events;
    using DevExpress.DevAV.Chat.Model;
    using DevExpress.Mvvm;
    using DevExpress.Mvvm.DataAnnotations;
    using DevExpress.Mvvm.POCO;
    using DXHtmlMessengerSample.Services;

    public class MessengerViewModel {
        public virtual string Title {
            get;
            protected set;
        }
        IDispatcherService dispatcher;
        IChannel channel;
        public async Task OnLoad() {
            dispatcher = this.GetRequiredService<IDispatcherService>();
            // Load settings and initialize
            var settingsService = this.GetRequiredService<ISettingsService>();
            var theme = settingsService.Theme ?? "Light";
            var currentUser = settingsService.CurrentUser ?? "John Heart";
            // Open messenger channel
            var messageServer = this.GetRequiredService<IMessageServer>();
            channel = await messageServer.Create(currentUser);
            channel.Subscribe(OnChannelEvent);
            // Pass the channel into dependent ViewModels
            Messenger.Default.Send(channel);
            await dispatcher.BeginInvoke(() => Title = $"DX HTML MESSENGER (CS) - [{currentUser.ToUpper()}]");
        }
        int authCounter = 0;
        void OnChannelEvent(ChannelEvent @event) {
            var credentialsRequired = @event as CredentialsRequiredEvent;
            if(credentialsRequired != null) {
                if(0 == authCounter++) {
                    // provide the access token from local cache without interaction
                    var cacheQuery = QueryAccessTokenFromLocalAuthCache(@event.UserName, credentialsRequired.Salt);
                    credentialsRequired.SetAccessTokenQuery(cacheQuery);
                }
                else {
                    // or query access-token asynchronously for the specific user
                    var userQuery = QueryAccessTokenFromUser(@event.UserName, credentialsRequired.Salt);
                    credentialsRequired.SetAccessTokenQuery(userQuery);
                }
            }
        }
        Task<string> QueryAccessTokenFromLocalAuthCache(string userName, string salt) {
            // simple emulation of local auth cache
            return Task.FromResult(DevAVEmpployeesInMemoryServer.GetPasswordHash(string.Empty, salt));
        }
        Task<string> QueryAccessTokenFromUser(string userName, string salt) {
            var accessTokenQueryCompletionSource = new TaskCompletionSource<string>();
            dispatcher.BeginInvoke(() => {
                var signInViewModel = SignInViewModel.Create(userName, salt);
                signInViewModel.ShowDialog();
                if(!string.IsNullOrEmpty(signInViewModel.AccessToken))
                    accessTokenQueryCompletionSource.SetResult(signInViewModel.AccessToken);
                else
                    accessTokenQueryCompletionSource.SetCanceled();
            });
            return accessTokenQueryCompletionSource.Task;
        }
        public void OnClosed() {
            if(channel != null)
                channel.Dispose();
            channel = null;
        }
        public void LogOff() {
            if(channel != null)
                channel.Send(new LogOff(channel));
        }
        UserViewModel userViewModel;
        [Command(isCommand: false)]
        public void ShowUserInfo(UserInfo userInfo) {
            ShowPopup(userViewModel ?? (userViewModel = UserViewModel.Create(userInfo)), userInfo);
        }
        ContactViewModel contactViewModel;
        [Command(isCommand: false)]
        public void ShowContactInfo(UserInfo contactInfo) {
            ShowPopup(contactViewModel ?? (contactViewModel = ContactViewModel.Create(contactInfo)), contactInfo);
        }
        void ShowPopup(UserInfoViewModel viewModel, UserInfo info) {
            viewModel.SetParentViewModel(this);
            var popup = this.GetService<IWindowService>(viewModel.ServiceKey);
            popup.Show(null, viewModel, info, this);
        }
        public static void ShowUserInfo(object viewModel, UserInfo userInfo) {
            var messenger = viewModel.GetParentViewModel<MessengerViewModel>();
            if(messenger != null)
                messenger.ShowUserInfo(userInfo);
        }
        public static void ShowContactInfo(object viewModel, UserInfo contactInfo) {
            var messenger = viewModel.GetParentViewModel<MessengerViewModel>();
            if(messenger != null)
                messenger.ShowContactInfo(contactInfo);
        }
    }
}

File: VB/ViewModels/MessengerViewModel.vb
Imports System.Threading.Tasks
Imports DevExpress.DevAV.Chat
Imports DevExpress.DevAV.Chat.Commands
Imports DevExpress.DevAV.Chat.Events
Imports DevExpress.DevAV.Chat.Model
Imports DevExpress.Mvvm
Imports DevExpress.Mvvm.DataAnnotations
Imports DevExpress.Mvvm.POCO
Imports DXHtmlMessengerSample.Services

Namespace DXHtmlMessengerSample.ViewModels
    Public Class MessengerViewModel
        Public Overridable Property Title() As String
        Private dispatcher As IDispatcherService
        Private channel As IChannel
        Public Async Function OnLoad() As Task
            dispatcher = Me.GetRequiredService(Of IDispatcherService)()
            ' Load settings and initialize
            Dim settingsService = Me.GetRequiredService(Of ISettingsService)()
            Dim theme = If(settingsService.Theme, "Light")
            Dim currentUser = If(settingsService.CurrentUser, "John Heart")
            ' Open messenger channel
            Dim messageServer = Me.GetRequiredService(Of IMessageServer)()
            channel = Await messageServer.Create(currentUser)
            channel.Subscribe(AddressOf OnChannelEvent)
            ' Pass the channel into dependent ViewModels
            Messenger.Default.Send(channel)
            Await dispatcher.BeginInvoke(Sub() Title = $"DX HTML MESSENGER (VB) - [{currentUser.ToUpper()}]")
        End Function
        Private AuthCounter As Integer = 0
        Sub OnChannelEvent(ByVal [event] As ChannelEvent)
            Dim credentialsRequired = TryCast([event], CredentialsRequiredEvent)
            If credentialsRequired IsNot Nothing Then
                AuthCounter += 1
                If AuthCounter = 1 Then
                    ' provide the access token from local cache without interaction
                    Dim cacheQuery = QueryAccessTokenFromLocalAuthCache([event].UserName, credentialsRequired.Salt)
                    credentialsRequired.SetAccessTokenQuery(cacheQuery)
                Else
                    ' or query access-token asynchronously for the specific user
                    Dim userQuery = QueryAccessTokenFromUser([event].UserName, credentialsRequired.Salt)
                    credentialsRequired.SetAccessTokenQuery(userQuery)
                End If
            End If
        End Sub
        Function QueryAccessTokenFromLocalAuthCache(ByVal userName As String, ByVal salt As String) As Task(Of String)
            ' simple emulation of local auth cache
            Return Task.FromResult(DevAVEmpployeesInMemoryServer.GetPasswordHash(String.Empty, salt))
        End Function
        Function QueryAccessTokenFromUser(ByVal userName As String, ByVal salt As String) As Task(Of String)
            Dim accessTokenQueryCompletionSource = New TaskCompletionSource(Of String)()
            dispatcher.BeginInvoke(Sub()
                                       Dim signInViewModel = ViewModels.SignInViewModel.Create(userName, salt)
                                       signInViewModel.ShowDialog()
                                       If Not String.IsNullOrEmpty(signInViewModel.AccessToken) Then
                                           accessTokenQueryCompletionSource.SetResult(signInViewModel.AccessToken)
                                       Else
                                           accessTokenQueryCompletionSource.SetCanceled()
                                       End If
                                   End Sub)
            Return accessTokenQueryCompletionSource.Task
        End Function
        Public Sub OnClosed()
            If channel IsNot Nothing Then
                channel.Dispose()
            End If
            channel = Nothing
        End Sub
        Public Sub LogOff()
            If channel IsNot Nothing Then
                channel.Send(New LogOff(channel))
            End If
        End Sub
        Private userViewModelCore As UserViewModel
        <Command(False)>
        Public Sub ShowUserInfo(ByVal userInfo As UserInfo)
            If userViewModelCore IsNot Nothing Then
                ShowPopup(userViewModelCore, userInfo)
            Else
                userViewModelCore = ViewModels.UserViewModel.Create(userInfo)
                ShowPopup(userViewModelCore, userInfo)
            End If
        End Sub
        Private contactViewModelCore As ContactViewModel
        <Command(False)>
        Public Sub ShowContactInfo(ByVal contactInfo As UserInfo)
            If contactViewModelCore IsNot Nothing Then
                ShowPopup(contactViewModelCore, contactInfo)
            Else
                contactViewModelCore = ViewModels.ContactViewModel.Create(contactInfo)
                ShowPopup(contactViewModelCore, contactInfo)
            End If
        End Sub
        Sub ShowPopup(ByVal viewModel As UserInfoViewModel, ByVal info As UserInfo)
            viewModel.SetParentViewModel(Me)
            Dim popup = Me.GetService(Of IWindowService)(viewModel.ServiceKey)
            popup.Show(Nothing, viewModel, info, Me)
        End Sub
        Public Shared Sub ShowUserInfo(ByVal viewModel As Object, ByVal userInfo As UserInfo)
            Dim messenger = POCOViewModelExtensions.GetParentViewModel(Of MessengerViewModel)(viewModel)
            If messenger IsNot Nothing Then
                messenger.ShowUserInfo(userInfo)
            End If
        End Sub
        Public Shared Sub ShowContactInfo(ByVal viewModel As Object, ByVal contactInfo As UserInfo)
            Dim messenger = POCOViewModelExtensions.GetParentViewModel(Of MessengerViewModel)(viewModel)
            If messenger IsNot Nothing Then
                messenger.ShowContactInfo(contactInfo)
            End If
        End Sub
    End Class
End Namespace